File: ObjC/SampleApp/RectangleViewController.m
///
/// @file RectangleViewController.m
/// @brief Implementation for RectangleViewController
///
/// @copyright Copyright Yahoo, Licensed under the terms of the Apache 2.0 license . See LICENSE file in project root for terms.
///

#import "RectangleViewController.h"
#import "AdPlacementConsts.h"
#import <YahooAds/YahooAds.h>

static const CGFloat kFadeInOutDuration = 0.3;

@interface RectangleViewController () <YASInlineAdViewDelegate>
@property (nonatomic, strong) IBOutlet UIButton *requestButton;
@property (nonatomic) IBOutlet UIActivityIndicatorView *spinnerView;
@property (nonatomic, strong) YASInlineAdView *inlineAd;
@end

@implementation RectangleViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.title = @"Inline Rectangle";
    [self requestInline:nil];
}

- (IBAction)requestInline:(id)sender
{
    [self destroyAd];
    [self uiFetchInProgress:YES];
    self.inlineAdContainer.alpha = 0.0;
    
    self.inlineAd = [[YASInlineAdView alloc] initWithPlacementId:kSampleAppRectangleAdPlacementID];
    self.inlineAd.delegate = self;
    [self.inlineAd loadWithPlacementConfig:nil];
}

- (void)showAd
{
    self.inlineAdContainer.alpha = 0.0;
    self.inlineAd.frame = self.inlineAdContainer.bounds;
    [self.inlineAdContainer addSubview:self.inlineAd];
    
    [UIView animateWithDuration:kFadeInOutDuration animations:^{
        self.inlineAdContainer.alpha = 1.0;
    } completion: nil];
}

- (void)destroyAd
{
    [self.inlineAd removeFromSuperview];
    self.inlineAd = nil;
}

- (void)uiFetchInProgress:(BOOL)inProgress
{
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        __strong typeof(self) strongSelf = weakSelf;
        if (inProgress) {
            [strongSelf.spinnerView startAnimating];
        }
        else {
            [strongSelf.spinnerView stopAnimating];
        }
        strongSelf.requestButton.enabled = ! inProgress;
    });
}

#pragma mark - YASInlineAdViewDelegate

- (void)inlineAdDidLoad:(YASInlineAdView *)inlineAd
{
    NSLog(@"Inline ad did load");
    self.inlineAd = inlineAd;
    
    [self uiFetchInProgress:NO];
    
    __weak typeof(self) weakSelf = self;
    dispatch_async(dispatch_get_main_queue(), ^{
        __strong typeof(self) strongSelf = weakSelf;
        if (strongSelf == nil) {
            return;
        }
        [strongSelf showAd];
    });
}

- (void)inlineAdLoadDidFail:(YASInlineAdView *)inlineAd withError:(YASErrorInfo *)errorInfo
{
    NSLog(@"Inline ad load failed to with error: %@", errorInfo);
    [self uiFetchInProgress:NO];
}

- (void)inlineAdDidFail:(YASInlineAdView *)inlineAd withError:(YASErrorInfo *)errorInfo
{
    NSLog(@"Inline ad failed to display with error: %@", errorInfo);
}

- (void)inlineAdDidExpand:(YASInlineAdView *)inlineAd
{
    NSLog(@"Inline ad expanded");
}

- (void)inlineAdDidCollapse:(YASInlineAdView *)inlineAd
{
    NSLog(@"Inline ad collapsed");
}

- (void)inlineAdClicked:(YASInlineAdView *)inlineAd
{
    NSLog(@"Inline ad clicked");
    [YahooAudiences logEventWithEventType:YahooAudiencesEventAdClick andParamBuilder:nil];
}

- (void)inlineAdDidLeaveApplication:(YASInlineAdView *)inlineAd
{
    NSLog(@"Inline ad did cause user to leave application");
}

- (void)inlineAdDidResize:(YASInlineAdView *)inlineAd
{
    NSLog(@"Inline ad did resize");
}

- (void)inlineAdDidRefresh:(YASInlineAdView *)inlineAd
{
    NSLog(@"Inline ad did refresh");
}

- (nullable UIViewController *)inlineAdPresentingViewController
{
    return self;
}

- (void)inlineAd:(nonnull YASInlineAdView *)inlineAd event:(nonnull NSString *)eventId source:(nonnull NSString *)source arguments:(nonnull NSDictionary<NSString *,id> *)arguments
{
     NSLog(@"Inline ad event occurred with ID: %@", eventId);
}

@end

File: Swift/SampleApp/RectangleViewController.swift
///
/// @file RectangleViewController.swift
/// @brief Implementation for RectangleViewController
///
/// @copyright Copyright Yahoo, Licensed under the terms of the Apache 2.0 license . See LICENSE file in project root for terms.
///

import Foundation
import UIKit
import YahooAds

class RectangleViewController : UIViewController {
    
    // MARK: - Properties
    
    @IBOutlet weak var adContainer: UIView?
    @IBOutlet weak var requestAdButton : UIButton?
    @IBOutlet weak var spinnerView: UIActivityIndicatorView?

    private var ad: YASInlineAdView?
    private let fadeInOutDuration = 0.3
    
    // MARK: - Life cycle
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        requestInline(self)
    }
    
    // MARK: User Interaction
    
    @IBAction func requestInline(_ sender: AnyObject) {
        destroy()

        spinnerView?.startAnimating()
        adContainer?.alpha = 0.0
        requestAdButton?.isEnabled = false
        
        ad =  YASInlineAdView(placementId: AdPlacementData.rectangleId)
        ad?.delegate = self
        ad?.load(with: nil)
    }
    
    func destroy() {
        ad?.removeFromSuperview()
        ad = nil
    }
}

// MARK: - YASInlineAdViewDelegate

extension RectangleViewController: YASInlineAdViewDelegate {
    
    func inlineAdDidLoad(_ inlineAd: YASInlineAdView) {
        print("Inline ad did load")
        ad = inlineAd
        
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.spinnerView?.stopAnimating()
            if let container = self.adContainer {
                inlineAd.frame = container.bounds
                container.addSubview(inlineAd)
            }
            UIView.animate(withDuration: self.fadeInOutDuration) { [weak self] in
                self?.adContainer?.alpha = 1.0
                self?.requestAdButton?.isEnabled = true
            }
        }
    }
    
    func inlineAdLoadDidFail(_ inlineAd: YASInlineAdView, withError errorInfo: YASErrorInfo) {
        print("Inline ad load failed with error:  \(errorInfo)")
        ad = nil
        
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            self.spinnerView?.stopAnimating()
            UIView.animate(withDuration: self.fadeInOutDuration) { [weak self] in
                self?.adContainer?.alpha = 1.0
                self?.requestAdButton?.isEnabled = true
            }
        }
    }
    
    func inlineAdPresentingViewController() -> UIViewController? {
        return self
    }

    func inlineAd(_ inlineAd: YASInlineAdView, event eventId: String, source: String, arguments: [String : Any]) {
        print("Inline ad event occurred with ID: \(eventId)")
    }

    func inlineAdDidFail(_ inlineAd: YASInlineAdView, withError errorInfo: YASErrorInfo) {
        print("Inline ad failed to display with error:  \(errorInfo)")
    }
    
    func inlineAdDidExpand(_ inlineAd: YASInlineAdView) {
        print("Inline ad expanded")
    }
    
    func inlineAdDidCollapse(_ inlineAd: YASInlineAdView) {
        print("Inline ad collapsed")
    }
    
    func inlineAdClicked(_ inlineAd: YASInlineAdView) {
        print("Inline ad clicked")
        
        YahooAudiences.logEvent(YahooAudiencesEvent.adClick, with: nil)
    }
    
    func inlineAdDidLeaveApplication(_ inlineAd: YASInlineAdView) {
        print("Inline ad did cause user to leave application")
    }
    
    func inlineAdDidResize(_ inlineAd: YASInlineAdView) {
        print("Inline ad did resize")
    }
    
    func inlineAdDidRefresh(_ inlineAd: YASInlineAdView) {
        print("Inline ad did refresh")
    }
}
